torchdata/dataloader2/utils/dispatch.py

```diff
@@ -0,0 +1,122 @@
+# LICENSE file in the root directory of this source tree.
+
+from multiprocessing.queues import Queue
+from typing import Dict, List, Optional, Set
+
+from torchdata.dataloader2.graph import DataPipe, DataPipeGraph, list_dps, traverse_dps
+
+from torchdata.datapipes.iter import IterDataPipe, ShardingRoundRobinDispatcher
+
+
+__all__ = ["_DummyIterDataPipe", "find_lca_non_replicable_dp", "find_replicable_branches"]
+
+
+class _DummyIterDataPipe(IterDataPipe):
+    r"""
+    This DataPipe is a placeholder to be replaced by the ``QueueWrapper``
+    that connects the worker process for non-replicable DataPipe.
+    """
+    req_queue: Queue
+    res_queue: Queue
+
+
+def find_lca_non_replicable_dp(graph: DataPipeGraph) -> Optional[DataPipe]:
+    r"""
+    Given the graph of DataPipe generated by ``traverse_dps`` function, return the
+    DataPipe instance that is the lowest common ancestor of all non-replicable DataPipes
+    (all parent DataPipes of ``sharding_round_robin_dispatch``)
+
+    Note:
+      - If multiple branches share the same source DataPipe and any branch contains a
+        non-replicable DataPipe, the lowest common ancestor of all branches is returned.
+      - If there is any non-replicable DataPipe in a circular-referenced (sub)graph, the
+        whole (sub)graph is treated as non-replicable and the last DataPipe is returned.
+    """
+    assert len(graph) == 1, "DataPipeGraph should only contain a single output DataPipe"
+
+    def _is_round_robin_sharding(dp: DataPipe) -> bool:
+        return type(dp) == ShardingRoundRobinDispatcher
+
+    dps = list_dps(graph)
+    non_replicable_dps: Set[int] = set()
+    for dp in dps:
+        # Skip when it has been visited
+        if id(dp) in non_replicable_dps:
+            continue
+        if _is_round_robin_sharding(dp):
+            parent_dps = list_dps(traverse_dps(dp))
+            for par_dp in parent_dps:
+                non_replicable_dps.add(id(par_dp))
+
+    root_dp_id = list(graph.keys())[0]
+    root_dp, root_graph = graph[root_dp_id]
+
+    cache: Dict[int, Optional[DataPipe]] = {}
+
+    def helper(root_dp_id, root_dp, root_graph) -> Optional[DataPipe]:  # pyre-ignore
+        if root_dp_id in cache:
+            return cache[root_dp_id]
+        if root_dp_id in non_replicable_dps:
+            cache[root_dp_id] = root_dp
+            return root_dp
+        cache[root_dp_id] = None
+        non_replicable_parents = []
+        for dp_id, (dp, src_graph) in root_graph.items():
+            res = helper(dp_id, dp, src_graph)
+            if res is not None:
+                non_replicable_parents.append(res)
+        # `root_dp` becomes the lowest common ancestor of this branch,
+        # if there are more than one unique non-replicable DataPipe prior to it.
+        if len(non_replicable_parents) > 0:
+            # One unique non-replicable DataPipe
+            if len(non_replicable_parents) == 1 or all(
```

Comment:
curious: can we just do something like 
```python
if len(set(non_replicable_parents)) == 1:
    ...
```
?

Reply:
There can be case that multiple branches share the same non-replicable DataPipe. Then, the LCA should not be the new joined DataPipe. Example:
```mermaid
graph TD;
    DP0["non-replicable DP0"]-->DP1-->DP3-->DP4;
    DP0-->DP2-->DP4;
```
For `DP4`, `len(non_replicable_parents)==2` but both `non_replicable_parents` are the same `DP0`. Then, the LCA should be `DP0` rather than `DP4`.

Reply:
but in that case won't `all(dp == non_replicable_parents[0] for dp in non_replicable_parents)` also returns true?

Reply:
discussed offline. `DataPipe` can be unhashable so cannot use `set`